plugins/inputs/execd/shim/goshim_notwindows.go

```diff
@@ -0,0 +1,14 @@
+// +build !windows
+
+package shim
+
+import (
+	"os"
+	"os/signal"
+	"syscall"
+)
+
+func listenForCollectMetricsSignals(collectMetricsPrompt chan os.Signal) {
+	// just listen to all the signals.
+	signal.Notify(collectMetricsPrompt, syscall.SIGHUP, syscall.SIGUSR1, syscall.SIGUSR2)
+}
```

Comment:
I don't think I like that execd in general (not just the shim) can signal collection through unix signals, whether it's HUP or USR1 or USR2.  They're tricky to use in a simple program in most languages because they make read() and write() return EINTR.  If you're writing your data too slowly and you're in the middle of a write when there's another collection signal, you may lose data.  I'm not sure how go handles this, it might insulate the coder from EINTR but execd is meant to be used with any language.

If "newline on stdin" works to signal collection, and it works on posix and windows, and it's safer/easier to implement than unix async signals, why isn't "newline on stdin" the only way to signal collection?

Reply:
This is news to me. Looks like Go handles it for you: [example](https://go-review.googlesource.com/c/go/+/14484/), so definitely the shim would as well. I've used signals pretty extensively in Ruby as well and never run into EINTR, so hopefully it's common for languages to handle this.

Personally I'd prefer to recommend use of STDIN, but this prevents the user from also triggering their own data collection, as they can't write to Telegraf's STDIN pipe that shared with the process, so there may be some cases where signals are preferred because they can be sent from another process.

Reply:
> this prevents the user from also triggering their own data collection, as they can't write to Telegraf's STDIN pipe that shared with the process

I don't follow.  The subprocess can write a new metric to its stdout whenever it wants to.  Just because telegraf is signaling a new data collection interval with a newline on stdin doesn't mean the subprocess can't also listen for unix signals from other sources or do any other IPC to trigger collection.

Anyway, I understand the shim needs to follow the execd plugin design and this PR isn't a good place to rehash that design.  I missed that review back in Feb.  It would have been nice if I had brought this up then :)